File: ios/lib/features/scenario/models/scenario.dart
import 'scenario_content.dart';

/// Scenario version summary for history list
class ScenarioVersionSummary {
  final String id;
  final int version;
  final String userPrompt;
  final DateTime createdAt;

  const ScenarioVersionSummary({
    required this.id,
    required this.version,
    required this.userPrompt,
    required this.createdAt,
  });

  factory ScenarioVersionSummary.fromJson(Map<String, dynamic> json) {
    return ScenarioVersionSummary(
      id: json['id'] as String,
      version: (json['version'] as num).toInt(),
      userPrompt: json['user_prompt'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'version': version,
      'user_prompt': userPrompt,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

/// Complete scenario version with content
class ScenarioVersion {
  final String id;
  final int version;
  final ScenarioContent content;
  final DateTime createdAt;
  final List<String>? validationErrors;

  const ScenarioVersion({
    required this.id,
    required this.version,
    required this.content,
    required this.createdAt,
    this.validationErrors,
  });

  factory ScenarioVersion.fromJson(Map<String, dynamic> json) {
    return ScenarioVersion(
      id: json['id'] as String,
      version: (json['version'] as num).toInt(),
      content: ScenarioContent.fromJson(json['content'] as Map<String, dynamic>),
      createdAt: DateTime.parse(json['created_at'] as String),
      validationErrors: (json['validation_errors'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'version': version,
      'content': content.toJson(),
      'created_at': createdAt.toIso8601String(),
      if (validationErrors != null) 'validation_errors': validationErrors,
    };
  }
}

/// Scenario entity
class Scenario {
  final String id;
  final String title;
  final String status;
  final DateTime createdAt;
  final String? currentVersionId;
  final ScenarioVersion? currentVersion;

  const Scenario({
    required this.id,
    required this.title,
    required this.status,
    required this.createdAt,
    this.currentVersionId,
    this.currentVersion,
  });

  factory Scenario.fromJson(Map<String, dynamic> json) {
    return Scenario(
      id: json['id'] as String,
      title: json['title'] as String,
      status: json['status'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      currentVersionId: json['current_version_id'] as String?,
      currentVersion: json['current_version'] != null
          ? ScenarioVersion.fromJson(json['current_version'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'status': status,
      'created_at': createdAt.toIso8601String(),
      if (currentVersionId != null) 'current_version_id': currentVersionId,
      if (currentVersion != null) 'current_version': currentVersion!.toJson(),
    };
  }
}

/// Request to create a new scenario
class CreateScenarioRequest {
  final String description;

  const CreateScenarioRequest({
    required this.description,
  });

  factory CreateScenarioRequest.fromJson(Map<String, dynamic> json) {
    return CreateScenarioRequest(
      description: json['description'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'description': description,
    };
  }
}

/// Request to refine an existing scenario
class RefineScenarioRequest {
  final String prompt;

  const RefineScenarioRequest({
    required this.prompt,
  });

  factory RefineScenarioRequest.fromJson(Map<String, dynamic> json) {
    return RefineScenarioRequest(
      prompt: json['prompt'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'prompt': prompt,
    };
  }
}

File: backend/src/schemas/scenario.py
"""Scenario Pydantic schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class Condition(BaseModel):
    """Schema for a condition."""
    condition: str = Field(..., description="Machine-readable condition ID")
    description: str = Field(..., description="Human-readable description")


class ScenarioScene(BaseModel):
    """Schema for a scene within an act."""
    id: str = Field(..., description="Unique scene identifier")
    name: str = Field(..., description="Human-readable scene name")
    mandatory: bool = Field(...)
    description_for_ai: str = Field(...)
    dm_hints: list[str] = Field(...)
    possible_outcomes: list[str] = Field(...)


class ScenarioAct(BaseModel):
    """Schema for scenario act."""
    id: str = Field(..., description="Unique act identifier")
    name: str = Field(..., description="Human-readable act name")
    entry_condition: Condition = Field(..., description="Condition to enter this act")
    exit_conditions: list[Condition] = Field(..., description="Conditions to exit this act")
    scenes: list[ScenarioScene] = Field(..., description="Scenes in this act")


class ScenarioNPC(BaseModel):
    """Schema for scenario NPC."""

    id: str = Field(..., description="Unique NPC identifier")
    name: str = Field(..., description="NPC name")
    role: str = Field(..., description="NPC role (ally, enemy, neutral, quest_giver, etc.)")
    personality: str = Field(..., description="NPC personality traits")
    speech_style: str = Field(..., description="How the NPC speaks")
    secrets: list[str] = Field(default_factory=list, description="NPC's secrets")
    motivation: str = Field(..., description="NPC's primary motivation")


class ScenarioLocation(BaseModel):
    """Schema for scenario location."""

    id: str = Field(..., description="Unique location identifier")
    name: str = Field(..., description="Location name")
    atmosphere: str = Field(..., description="Location atmosphere and mood")
    rooms: list[str] = Field(default_factory=list, description="Rooms or areas in this location")


class ScenarioFlag(BaseModel):
    """Schema for scenario flag definition."""

    id: str = Field(..., description="Unique flag identifier (machine-readable key)")
    name: str = Field(..., description="Human-readable flag name for display")
    description: str = Field(..., description="Description of what this flag represents")


class ScenarioContent(BaseModel):
    """Schema for scenario content structure."""

    tone: str = Field(
        ...,
        description="Scenario tone (dark_fantasy, heroic, horror, mystery)",
    )
    difficulty: str = Field(
        ...,
        description="Difficulty level (beginner, intermediate, hardcore)",
    )
    players_min: int = Field(..., ge=1, le=5, description="Minimum number of players (1 for single player)")
    players_max: int = Field(..., ge=1, le=5, description="Maximum number of players")
    world_lore: str = Field(..., description="World background and lore")
    acts: list[ScenarioAct] = Field(..., description="Story acts")
    npcs: list[ScenarioNPC] = Field(..., description="Non-player characters")
    locations: list[ScenarioLocation] = Field(..., description="Game locations")
    flags: list[ScenarioFlag] = Field(default_factory=list, description="World flag definitions")


class CreateScenarioRequest(BaseModel):
    """Schema for creating a scenario."""

    description: str = Field(
        ...,
        min_length=10,
        max_length=2000,
        description="Natural language description of the desired scenario",
    )


class RefineScenarioRequest(BaseModel):
    """Schema for refining an existing scenario."""

    prompt: str = Field(
        ...,
        min_length=5,
        max_length=1000,
        description="Refinement instructions",
    )


class ScenarioVersionSummary(BaseModel):
    """Schema for scenario version summary."""

    id: UUID
    version: int
    user_prompt: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ScenarioVersionResponse(BaseModel):
    """Schema for scenario version with full content."""

    id: UUID
    version: int
    content: dict[str, Any]
    validation_errors: list[str] | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ScenarioResponse(BaseModel):
    """Schema for scenario response (without version details)."""

    id: UUID
    title: str
    status: str
    current_version_id: UUID | None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ScenarioWithVersionResponse(BaseModel):
    """Schema for scenario with current version details."""

    id: UUID
    title: str
    status: str
    current_version_id: UUID | None
    created_at: datetime
    current_version: ScenarioVersionResponse | None

    model_config = ConfigDict(from_attributes=True)
